api/v1beta1/flinkcluster_validate.go

```diff
@@ -183,7 +197,44 @@ func (v *Validator) checkSavepointGeneration(
 	// Check if only `savepointGeneration` changed, no other changes.
 	var oldCopy = old.DeepCopy()
 	oldCopy.Spec.Job.SavepointGeneration = newSpecGen
-	return reflect.DeepEqual(new.Spec, oldCopy.Spec), nil
+	if reflect.DeepEqual(new.Spec, oldCopy.Spec) {
+		return true, nil
+	}
+
+	return false, fmt.Errorf(
+		"you cannot update savepointGeneration with others at the same time")
+}
+
+// Check job update constraints.
+func (v *Validator) validateJobUpdate(
+	old *FlinkCluster, new *FlinkCluster) (bool, error) {
+	if old.Spec.Job == nil || new.Spec.Job == nil {
+		return false, nil
+	}
+	if !reflect.DeepEqual(old.Spec.Job.FromSavepoint, new.Spec.Job.FromSavepoint) {
+		return false, fmt.Errorf(
+			"updating fromSavepoint is not allowed")
+	}
+	if old.Spec.Job.SavepointsDir != nil && *old.Spec.Job.SavepointsDir != "" &&
+		(new.Spec.Job.SavepointsDir == nil || *new.Spec.Job.SavepointsDir == "") {
+		return false, fmt.Errorf(
+			"removing savepointsDir is not allowed")
+	}
+	if !reflect.DeepEqual(old.Spec.Job, new.Spec.Job) {
+		if old.Spec.Job.SavepointsDir == nil || *old.Spec.Job.SavepointsDir == "" {
+			return false, fmt.Errorf(
+				"updating job is not allowed when spec.job.savepointsDir was not provided")
+		}
+		oldCopy := old.DeepCopy()
+		oldCopy.Spec.Job = new.Spec.Job.DeepCopy()
+		if !reflect.DeepEqual(oldCopy.Spec, new.Spec) {
+			return false, fmt.Errorf(
+				"you cannot update fields other than spec.job, old FlinkCluster spec: %v, new FlinkCluster spec: %v", oldCopy.Spec, new.Spec)
```

Comment:
Did you consider allowing update Flink image? In that case, we might need to delete the cluster then recreate. For job update, this PR reused the existing cluster if available, but if we want to support both, we might want to use the same mechanism.

Reply:
I hope the Flink cluster update is also supported. However, I think it will be a little difficult to include in this PR because it seems to have to consider not only the Flink image but also the update of other fields in the JM / TM spec. I hope the overall cluster update feature is also supported.

Reply:
Right, if we support updating JM / TM spec, that would be more complete, e.g., many users are interested in scaling. If asking you to consider is that, maybe we should choose to always delete and recreate cluster when implementing job update in this PR. In that way, we could reuse the same mechanism to implement updating the whole cluster. Otherwise, we would end up with 2 mechanisms. What do you think?

Reply:
I think it is basically right to create a new container as a container operation method.  However, in the case of flink cluster, it seems to be a disadvantage that it is difficult to use the local recovery function that will shorten the job resume time as well as the cluster bootstrap time.  What about providing a cluster restart option like restartOnJobUpdate?

Reply:
Option sounds good to me. My point is that we need to take image / JM / tm into consideration at this point.

Reply:
Ok, I'm going to look it a little further.

Reply:
Improved to support both option, recreating or updating cluster.